src/ir/auth_logic/auth_logic_ast_traversing_visitor_test.cc

```diff
@@ -209,21 +210,64 @@ TEST(AuthLogicAstTraversingVisitorTest, SimpleTraversalTest) {
 
   Program program1({}, {saysAssertion1, saysAssertion2}, {query1, query2});
 
+  const SaysAssertion* saysAssertion1ptr =
+    std::addressof(program1.says_assertions()[0]);
+  const SaysAssertion* saysAssertion2ptr =
+    std::addressof(program1.says_assertions()[1]);
+  const Query* query1ptr =
+    std::addressof(program1.queries()[0]);
+  const Query* query2ptr =
+    std::addressof(program1.queries()[1]);
+  const Assertion* assertion1sa1ptr =
+    std::addressof(saysAssertion1ptr->assertions()[0]);
+  const Assertion* assertion1sa2ptr =
+    std::addressof(saysAssertion2ptr->assertions()[0]);
+  const Assertion* assertion2ptr =
+    std::addressof(saysAssertion2ptr->assertions()[1]);
+  const Fact* fact1sa1ptr = std::addressof(
+    std::get<Fact>(assertion1sa1ptr->GetValue()));
+  const Fact* fact1sa2ptr = std::addressof(
+    std::get<Fact>(assertion1sa2ptr->GetValue()));
+  const Fact* fact2ptr = std::addressof(
+    std::get<Fact>(assertion2ptr->GetValue()));
+
+  const Principal* prinBAssertionptr=
+    std::addressof(fact2ptr->delegation_chain().front());
+  const Principal* prinBQueryptr =
+    std::addressof(query2ptr->fact().delegation_chain().front());
+
+
   TraversalOrderVisitor preorder_visitor(TraversalType::kPre);
   program1.Accept(preorder_visitor);
   EXPECT_THAT(
       preorder_visitor.nodes(),
       testing::ElementsAre(
-          program1.DebugPrint(), saysAssertion1.DebugPrint(),
-          prinA.DebugPrint(), assertion1.DebugPrint(), fact1.DebugPrint(),
-          baseFact1.DebugPrint(), saysAssertion2.DebugPrint(),
-          prinC.DebugPrint(), assertion1.DebugPrint(), fact1.DebugPrint(),
-          baseFact1.DebugPrint(), assertion2.DebugPrint(), fact2.DebugPrint(),
-          prinB.DebugPrint(), baseFact2.DebugPrint(), query1.DebugPrint(),
-          prinA.DebugPrint(), fact1.DebugPrint(), baseFact1.DebugPrint(),
-          query2.DebugPrint(), prinB.DebugPrint(), fact2.DebugPrint(),
-          prinB.DebugPrint(), baseFact2.DebugPrint()));
-
+          std::addressof(program1),
+          saysAssertion1ptr,
```

Comment:
@bgogul I am trying to fix this test to work without `DebugPrint`. This is an attempt at collecting the addresses of the ast nodes in the visitor rather than collecting `DebugPrint` strings (see line 185), but with the way `auth_logic::ast` is implemented, I am not sure there is a way to actually get the right addresses. And in any case this looks like it would be awful to update if the test needs to be updated.

Is there a better way to write this test?

The next thing I was thinking to try is to write a class `ASTNode` that has `==operator` as a pure virtual method and having both the `ast.h` and `program.h` classes extend from this. Then I think this visitor could use `std::vector<ASTNode` as the type of `node_` as long as `EXPECT_THAT` would then use the `==` operator in that case. Do you happen to know if that will/won't work or is there a better way to do this?

Reply:
@bgogul OK I tried using the abstract class idea instead, but it looks like this isn't compatible with `EXPECT_THAT` for some reason?

I get:
```
external/com_google_googletest/googletest/include/gtest/gtest-matchers.h:880:40: error: no matching function for call to 'Eq'
```

Any idea if I am just implementing this idea wrong or if this fundamentally won't work?

Reply:
Perhaps @markww might also know why `ASTNode` is not enough to make `auth_logic_ast_traversing_visitor_test` work with `EXPECT_THAT` and a `std::vector<ASTNode>`